grove/tomography/tomography.py

```diff
@@ -0,0 +1,727 @@
+
+        rho0 = ut.n_qubit_ground_state(nqc)
+
+        n_lkj = np.asarray(histograms)
+
+        b_jkl_mn = _prepare_b_jkl_mn(readout_povm, pauli_basis, pre_channel_ops,
+                                     post_channel_ops, rho0)
+
+        r_mn = cvxpy.Variable(pauli_basis.dim ** 2)
+        p_jkl = b_jkl_mn.real * r_mn
+        obj = -np.matrix(n_lkj.ravel()) * cvxpy.log(p_jkl)
+
+        # cvxpy has col-major order and we collapse k and l onto single dimension
+        p_jkl_mat = cvxpy.reshape(p_jkl, pi_basis.dim, len(pre_channel_ops) * len(post_channel_ops))
+
+        # Default constraints:
+        # MLE must describe valid probability distribution
+        # i.e., for each k and l, p_jkl must sum to one and be element-wise non-negative:
+        # 1. \sum_j p_jkl == 1  for all k, l
+        # 2. p_jkl >= 0         for all j, k, l
+        # where p_jkl = \sum_m b_jkl_mn r_mn
+        constraints = [p_jkl >= 0,
+                       np.matrix(np.ones((1, pi_basis.dim))) * p_jkl_mat == 1]
+        r_mn_mat = cvxpy.reshape(r_mn, pauli_basis.dim, pauli_basis.dim)
+        super_pauli_basis = pauli_basis.super_basis()
+        choi_real_imag = sum((r_mn_mat[jj, kk] * ut.to_realimag(
+            super_pauli_basis.ops[jj + kk * pauli_basis.dim])
+                              for jj in range(pauli_basis.dim)
+                              for kk in range(pauli_basis.dim)), 0)
+
+        if COMPLETELY_POSITIVE in settings.constraints:
+            if _SDP_SOLVER.is_functional():
+                constraints.append(choi_real_imag >> 0)
+            else:  # pragma no coverage
+                _log.warning("No convex solver capable of semi-definite problems installed.\n"
+                             "Dropping the complete positivity constraint on the process")
+
+        if TRACE_PRESERVING in settings.constraints:
+            constraints.append(r_mn_mat[0, 0] == 1)
+            constraints.append(r_mn_mat[0, 1:] == 0)
+
+        prob = cvxpy.Problem(cvxpy.Minimize(obj), constraints)
+        _ = prob.solve(solver=SOLVER, **settings.solver_kwargs)
+
+        r_mn_est = r_mn.value.reshape((pauli_basis.dim, pauli_basis.dim)).transpose()
+        return ProcessTomography(r_mn_est, pauli_basis, settings)
+
+    def __init__(self, r_est, pauli_basis, settings):
+        """
+        Construct a ProcessTomography to encapsulate the result of estimating a quantum process
+        from a quantum tomography measurement.
+
+        :param numpy.ndarray r_est: The estimated quantum process represented as a Pauli transfer
+         matrix.
+        :param OperatorBasis pauli_basis: The employed (generalized) Pauli basis.
+        :param TomographySettings settings: The settings used to estimate the process.
+        """
+        self.r_est = r_est
+        self.sop = pauli_basis.super_from_tm(r_est)
+        self.pauli_basis = pauli_basis
+        self.settings = settings
+
+    def process_fidelity(self, other):
```

Comment:
this signature is misleading to me. With other I associate an object of the same type as this class, i.e. a Tomogrpahy instance

Reply:
:+1: